File: disposia-engine/src/main/java/exo/engine/parse/rss/RomeFeedParser.java
package exo.engine.parse.rss;

import com.rometools.modules.atom.modules.AtomLinkModule;
import com.rometools.modules.content.ContentModule;
import com.rometools.modules.itunes.EntryInformation;
import com.rometools.modules.itunes.FeedInformation;
import com.rometools.modules.itunes.types.Category;
import com.rometools.rome.feed.atom.Link;
import com.rometools.rome.feed.module.Module;
import com.rometools.rome.feed.synd.SyndEnclosure;
import com.rometools.rome.feed.synd.SyndEntry;
import com.rometools.rome.feed.synd.SyndFeed;
import com.rometools.rome.feed.synd.SyndImage;
import com.rometools.rome.io.FeedException;
import com.rometools.rome.io.SyndFeedInput;
import exo.engine.domain.dto.*;
import exo.engine.exception.FeedParsingException;
import exo.engine.parse.rss.rome.PodloveSimpleChapterModule;
import exo.engine.util.UrlUtil;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.xml.sax.InputSource;

import java.io.StringReader;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.util.LinkedHashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

import static com.google.common.base.Strings.isNullOrEmpty;

/**
 * @author Maximilian Irro
 */
public class RomeFeedParser implements FeedParser {

    private static final Logger log = LoggerFactory.getLogger(RomeFeedParser.class);

    private final Podcast thePodcast;
    private final List<Episode> theEpisodes;

    private RomeFeedParser(String xmlData) throws FeedParsingException {
        try {
            final InputSource inputSource = new InputSource( new StringReader( xmlData ) );
            final SyndFeedInput input = new SyndFeedInput();
            final SyndFeed feed = input.build(inputSource);

            this.thePodcast = parseFeed(feed);
            this.theEpisodes = extractEpisodes(feed);
        } catch (FeedException | IllegalArgumentException e) {
            throw new FeedParsingException("RomeFeedParser could not parse the feed", e);
        }
    }

    public static RomeFeedParser of(String xmlData) throws FeedParsingException {
        return new RomeFeedParser(xmlData);
    }

    @Override
    public Podcast getPodcast() {
        return this.thePodcast;
    }

    @Override
    public List<Episode> getEpisodes() {
        return this.theEpisodes;
    }

    private Podcast parseFeed(SyndFeed feed) {
        final ImmutablePodcast.Builder builder = ImmutablePodcast.builder();
        String link = UrlUtil.sanitize(feed.getLink());

        builder
            .setTitle(feed.getTitle())
            .setLink(link)
            .setDescription(feed.getDescription());

        SyndImage img = null;
        if (feed.getImage() != null) {
            img = feed.getImage();
            if(!isNullOrEmpty(img.getUrl())){
                builder.setImage(img.getUrl());
            }

            // now, it title/link/description were NULL, we use the values set in
            // the image tag as fallbacks because they usually have the same values
            if (isNullOrEmpty(feed.getTitle()) && !isNullOrEmpty(img.getTitle())) {
                builder.setTitle(img.getTitle());
            }
            if (link == null && !isNullOrEmpty(img.getLink())) {
                link = UrlUtil.sanitize(img.getLink());
                builder.setLink(link);
            }
            if (isNullOrEmpty(feed.getDescription()) && !isNullOrEmpty(img.getDescription())) {
                builder.setDescription(img.getDescription());
            }
        }

        if (feed.getPublishedDate() != null) {
            builder.setPubDate(LocalDateTime.ofInstant(feed.getPublishedDate().toInstant(), ZoneId.systemDefault()));
        }
        builder
            .setLanguage(feed.getLanguage())
            .setGenerator(feed.getGenerator())
            .setCopyright(feed.getCopyright())
            .setDocs(feed.getDocs())
            .setManagingEditor(feed.getManagingEditor());

        // access the <itunes:...> entries
        final Module itunesFeedModule = feed.getModule(FeedInformation.URI);
        final FeedInformation itunes = (FeedInformation) itunesFeedModule;
        if (itunes != null) {
            builder
                .setItunesSummary(itunes.getSummary())
                .setItunesAuthor(itunes.getAuthor())
                .setItunesKeywords(String.join(", ", itunes.getKeywords()));

            // we set the itunes image as a fallback only
            if (itunes.getImage() != null) {
                if (img == null || isNullOrEmpty(img.getUrl())) {
                    builder.setImage(itunes.getImage().toExternalForm());
                }
            }
            builder
                .setItunesCategories(new LinkedHashSet<>(
                    itunes.getCategories().stream()
                        .map(Category::getName)
                        .collect(Collectors.toCollection(LinkedList::new))))
                .setItunesExplicit(itunes.getExplicit())
                .setItunesBlock(itunes.getBlock())
                .setItunesType(itunes.getType())
                .setItunesOwnerName(itunes.getOwnerName())
                .setItunesOwnerEmail(itunes.getOwnerEmailAddress());
            //builder.setItunesCategory(String.join(" | ", itunesFeedInfo.getCategories().stream().toImmutable(c->c.getName()).collect(Collectors.toCollection(LinkedList::new))));
        } else {
            log.debug("No iTunes Namespace elements found in Podcast");
        }

        // here I process the feed specific atom Links
        final List<Link> atomLinks = getAtomLinks(feed);
        for (Link atomLink : atomLinks) {
            if (atomLink.getRel().equals("http://podlove.org/deep-link")) {
                // TODO this should be a link to the episode website (but is it always though?!)
            } else if (atomLink.getRel().equals("payment")) {
                // TODO
            } else if (atomLink.getRel().equals("self")) {
                // TODO
            } else if (atomLink.getRel().equals("alternate")) {
                // TODO
            } else if (atomLink.getRel().equals("first")) {
                // TODO
            } else if (atomLink.getRel().equals("next")) {
                // TODO
            } else if (atomLink.getRel().equals("last")) {
                // TODO
            } else if (atomLink.getRel().equals("hub")) {
                // TODO
            } else if (atomLink.getRel().equals("search")) {
                // TODO
            } else if (atomLink.getRel().equals("via")) {
                // TODO
            } else if (atomLink.getRel().equals("related")) {
                // TODO
            } else if (atomLink.getRel().equals("prev-archive")) {
            } else {
                log.warn("Came across an <atom:link> with a relation I do not handle : '{}'", atomLink.getRel());
            }
        }

        return builder.create();
    }

    private List<Episode> extractEpisodes(SyndFeed feed) {
        final List<Episode> results = new LinkedList<>();
        for (SyndEntry e : feed.getEntries()) {
            final ImmutableEpisode.Builder builder = ImmutableEpisode.builder();

            builder
                .setTitle(e.getTitle())
                .setLink(UrlUtil.sanitize(e.getLink()));

            if (e.getPublishedDate() != null) {
                builder.setPubDate(LocalDateTime.ofInstant(e.getPublishedDate().toInstant(), ZoneId.systemDefault()));
            }
            //doc.setGuid(TODO);
            if (e.getDescription() != null) {
                builder.setDescription(e.getDescription().getValue());
            }

            if (e.getUri() != null) {
                builder.setGuid(e.getUri());
            }

            if (e.getEnclosures() != null && e.getEnclosures().size() > 0) {
                final SyndEnclosure enclosure = e.getEnclosures().get(0);
                builder
                    .setEnclosureUrl(enclosure.getUrl())
                    .setEnclosureType(enclosure.getType())
                    .setEnclosureLength(enclosure.getLength());
                if (e.getEnclosures().size() > 1) {
                    log.warn("Encountered multiple <enclosure> elements in <item> element");
                }
            }

            // access the <content:encoded> entries
            final Module contentModule = e.getModule(ContentModule.URI);
            final ContentModule content = (ContentModule) contentModule;
            if (content != null) {
                if (content.getEncodeds().size() > 0) {
                    builder.setContentEncoded(content.getEncodeds().get(0));
                    if (content.getEncodeds().size() > 1) {
                        log.warn("Encountered multiple <content:encoded> elements in <item> element");
                    }
                }
            }

            // access the <itunes:...> entries
            final Module itunesEntryModule = e.getModule(EntryInformation.URI);
            final EntryInformation itunes = (EntryInformation) itunesEntryModule;
            if (itunes != null) {
                if (itunes.getImage() != null) {
                    builder.setImage(itunes.getImage().toExternalForm());
                }
                if (itunes.getDuration() != null) {
                    builder.setItunesDuration(itunes.getDuration().toString());
                }
                builder
                    .setItunesSubtitle(itunes.getSubtitle())
                    .setItunesAuthor(itunes.getAuthor())
                    .setItunesSummary(itunes.getSummary())
                    .setItunesSeason(itunes.getSeason())
                    .setItunesEpisode(itunes.getEpisode())
                    .setItunesEpisodeType(itunes.getEpisodeType());
            } else {
                log.debug("No iTunes Namespace elements found in Episode");
            }

            // access the <psc:chapter> entries
            final Module pscEntryModule = e.getModule(PodloveSimpleChapterModule.URI);
            final PodloveSimpleChapterModule simpleChapters = ((PodloveSimpleChapterModule) pscEntryModule);
            if (simpleChapters != null) {
                if (simpleChapters.getChapters() != null && simpleChapters.getChapters().size() > 0) {
                    builder.setChapters(
                        simpleChapters.getChapters().stream()
                            .map(sc -> ImmutableChapter.builder()
                                .setStart(sc.getStart())
                                .setTitle(sc.getTitle())
                                .setHref(sc.getHref())
                                .create())
                            .collect(Collectors.toList()));
                }
            } else {
                log.debug("No Podlove Simple Chapter marks found in Episode");
            }

            results.add(builder.create());
        }

        return results;
    }

    private List<Link> getAtomLinks(SyndFeed syndFeed) {
        final Module atomFeedModule = syndFeed.getModule(AtomLinkModule.URI);
        final AtomLinkModule atomLinkModule = (AtomLinkModule) atomFeedModule;
        return Optional
            .ofNullable(atomLinkModule)
            .map(AtomLinkModule::getLinks)
            .orElse(new LinkedList<>());
    }

}

File: disposia-engine/src/main/scala/io/disposia/engine/parser/feed/RomeFeedParser.scala
package io.disposia.engine.parser.feed

import java.io.StringReader

import com.google.common.base.Strings.isNullOrEmpty
import com.rometools.modules.itunes.FeedInformation
import com.rometools.rome.feed.synd.{SyndEntry, SyndFeed, SyndImage}
import com.rometools.rome.io.SyndFeedInput
import com.typesafe.scalalogging.Logger
import io.disposia.engine.domain.info._
import io.disposia.engine.domain.{Chapter, Episode, Podcast}
import io.disposia.engine.parse.rss.RomeModuleExtractor
import io.disposia.engine.util.mapper.{DateMapper, UrlMapper}
import org.xml.sax.InputSource

import scala.collection.JavaConverters._
import scala.compat.java8.OptionConverters._


// TODO rewrite this class in idiomatic Scala
class RomeFeedParser(private val xmlData: String) {

  private val log: Logger = Logger(getClass)

  private val inputSource: InputSource = new InputSource(new StringReader(xmlData))
  private val input: SyndFeedInput = new SyndFeedInput
  private val feed: SyndFeed = input.build(inputSource)
  private val feedItunesModule: Option[FeedInformation] = RomeModuleExtractor.getItunesModule(feed).asScala

  val podcast: Podcast = parseFeed(feed)
  val episodes: List[Episode] = extractEpisodes(feed)

  /*
  try {

    // TODO

  } catch {
    case ex: Throwable =>
      log.error("RomeFeedParser could not parse the feed : {}", ex.getMessage)
      ex.printStackTrace()
  }
  */

  private def parseFeed(feed: SyndFeed): Podcast = {

    // # # # # # # # # # # # # # # # # # # # # # # #

    // TODO process the atomlinks
    RomeModuleExtractor.getAtomLinks(feed).asScala
      .foreach { atomLink =>
        if (atomLink.getRel == "http://podlove.org/deep-link") {
          // TODO this should be a link to the episode website (but is it always though?!)
        }
        else if (atomLink.getRel == "payment") {
          // TODO
        }
        else if (atomLink.getRel == "self") {
          // TODO
        }
        else if (atomLink.getRel == "alternate") {
          // TODO
        }
        else if (atomLink.getRel == "first") {
          // TODO
        }
        else if (atomLink.getRel == "next") {
          // TODO
        }
        else if (atomLink.getRel == "last") {
          // TODO
        }
        else if (atomLink.getRel == "hub") {
          // TODO
        }
        else if (atomLink.getRel == "search") {
          // TODO
        }
        else if (atomLink.getRel == "via") {
          // TODO
        }
        else if (atomLink.getRel == "related") {
          // TODO
        }
        else if (atomLink.getRel == "prev-archive") {
          // TODO
        }
        else log.warn("Came across an <atom:link> with a relation I do not handle : '{}'", atomLink.getRel)
      }

    Podcast(
      id              = None,
      title           = podcastTitleWithImageFallback,
      link            = podcastLinkWithImageFallback,
      description     = podcastDescriptionWithImageFallback,
      pubDate         = DateMapper.asLocalDateTime(feed.getPublishedDate),
      image           = podcastImageWithItunesFallback,
      meta = PodcastMetadata(
        lastBuildDate  = None,  // TODO the parser does not yet produce this
        language       = Option(feed.getLanguage),
        generator      = Option(feed.getGenerator),
        copyright      = Option(feed.getCopyright),
        docs           = Option(feed.getDocs),
        managingEditor = Option(feed.getManagingEditor),
      ),
      registration = PodcastRegistrationInfo(
        timestamp = None,
        complete  = None,
      ),
      itunes    = podcastItunesInfo,
      feedpress = podcastFeedpressInfo,
      fyyd      = podcastFyydInfo,
    )
  }

  private def extractEpisodes(feed: SyndFeed): List[Episode] = feed
    .getEntries.asScala
    .map(e => extractEpisode(e))
    .toList

  private def extractEpisode(e: SyndEntry): Episode = Episode(
    id              = None,
    podcastId       = None,
    podcastTitle    = podcast.title,
    title           = Option(e.getTitle),
    link            = UrlMapper.sanitize(e.getLink),
    pubDate         = DateMapper.asLocalDateTime(e.getPublishedDate),
    guid            = Option(e.getUri),
    guidIsPermalink = None, // TODO welches feld von ROME korrespondiert zu diesem Boolean?
    description     = Option(e.getDescription).map(_.getValue),
    image           = episodeImage(e),
    contentEncoded  = episodeContentEncoded(e),
    chapters        = episodeChapters(e),
    itunes          = episodeItunesInfo(e),
    enclosure       = episodeEnclosureInfo(e),
    registration = EpisodeRegistrationInfo(
      timestamp = None,
    )
  )

  private def podcastTitleWithImageFallback: Option[String] = {
    val feedTitle = Option(feed.getTitle)
    val imgTitle = Option(feed.getImage).map(_.getTitle)

    (feedTitle, imgTitle) match {
      case (Some(_), _)    => feedTitle
      case (None, Some(_)) => imgTitle
      case (_, _)          => None
    }
  }

  private def podcastLinkWithImageFallback: Option[String] = {
    val feedLink = Option(feed.getLink)
    val imgLink = Option(feed.getImage).map(_.getLink)

    val link = (feedLink, imgLink) match {
      case (Some(_), _)    => feedLink
      case (None, Some(_)) => imgLink
      case (_, _)          => None
    }
    link.flatMap(UrlMapper.sanitize)
  }

  private def podcastDescriptionWithImageFallback: Option[String] = {
    val feedDescr = Option(feed.getDescription)
    val imgDescr = Option(feed.getImage).map(_.getDescription)

    (feedDescr, imgDescr) match {
      case (Some(_), _)    => feedDescr
      case (None, Some(_)) => imgDescr
      case (_, _)          => None
    }
  }

  private def podcastImageWithItunesFallback: Option[String] = {
    val feedImg: Option[String] = Option(feed.getImage)
      .flatMap { img: SyndImage =>
        if (isNullOrEmpty(img.getUrl))
          None
        else
          Some(img.getUrl)
      }

    val itunesImg: Option[String] = feedItunesModule
      .map(_.getImage)
      .map(_.toString)
      .orElse(None)

    (feedImg, itunesImg) match {
      case (Some(_), _)    => feedImg
      case (None, Some(_)) => itunesImg
      case (_, _)          => None
    }
  }

  private def podcastItunesInfo: PodcastItunesInfo = feedItunesModule
    .map { itunes =>
      PodcastItunesInfo(
        summary     = Option(itunes.getSummary),
        author      = Option(itunes.getAuthor),
        keywords    = Option(itunes.getKeywords)
          .map(_.toList)
          .getOrElse(Nil),
        categories  = Option(itunes.getCategories)
          .map(cs => cs.asScala.map(c => c.getName))
          .map(_.toSet)  // eliminate duplicates
          .map(_.toList)
          .getOrElse(Nil),
        explicit    = Option(itunes.getExplicit),
        block       = Option(itunes.getBlock),
        podcastType = Option(itunes.getType),
        ownerName   = Option(itunes.getOwnerName),
        ownerEmail  = Option(itunes.getOwnerEmailAddress),
      )
    }.getOrElse(PodcastItunesInfo())

  private def podcastFeedpressInfo: PodcastFeedpressInfo = PodcastFeedpressInfo(
    locale = None
  )

  private def podcastFyydInfo: PodcastFyydInfo = PodcastFyydInfo(
    verify = None
  )

  private def episodeImage(e: SyndEntry): Option[String] = RomeModuleExtractor
    .getItunesEntryInformation(e).asScala
    .flatMap { itunes =>
      Option(itunes.getImage).map(img => img.toExternalForm)
    }

  private def episodeItunesInfo(e: SyndEntry): EpisodeItunesInfo = RomeModuleExtractor
    .getItunesEntryInformation(e).asScala
    .map { itunes =>
      EpisodeItunesInfo(
        duration    = Option(itunes.getDuration).map(_.toString),
        subtitle    = Option(itunes.getSubtitle),
        author      = Option(itunes.getAuthor),
        summary     = Option(itunes.getSummary),
        season      = Option(itunes.getSeason),
        episode     = Option(itunes.getEpisode),
        episodeType = Option(itunes.getEpisodeType),
      )
    }.getOrElse(EpisodeItunesInfo())

  private def episodeEnclosureInfo(e: SyndEntry): EpisodeEnclosureInfo = Option(e.getEnclosures)
    .map { enclosures =>
      if (enclosures.size > 1) log.warn("Encountered multiple <enclosure> elements in <item> element")
      if (enclosures.size > 0) {
        val enclosure = e.getEnclosures.get(0)
        EpisodeEnclosureInfo(
          url    = Option(enclosure.getUrl),
          length = Option(enclosure.getLength),
          typ    = Option(enclosure.getType),
        )
      } else {
        EpisodeEnclosureInfo()
      }
    }.getOrElse(EpisodeEnclosureInfo())

  private def episodeContentEncoded(e: SyndEntry): Option[String] = RomeModuleExtractor
    .getContentModule(e).asScala
    .map { content =>
      if (content.getEncodeds.size > 1) log.warn("Encountered multiple <content:encoded> elements in <item> element")
      if (content.getEncodeds.size > 0) {
        content.getEncodeds.get(0)
      } else {
        null
      }
    }

  private def episodeChapters(e: SyndEntry): List[Chapter] = RomeModuleExtractor
    .getPodloveSimpleChapterModule(e).asScala
    .map(simpleChapters => simpleChapters.getChapters)
    .map { _.asScala
      .map { c =>
        Chapter(
          id        = None,
          episodeId = None,
          start     = Option(c.getStart),
          title     = Option(c.getTitle),
          href      = Option(c.getHref),
          image     = Option(c.getImage),
        )
      }.toList
    }.getOrElse(List())

}
